core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java

```diff
@@ -138,34 +153,32 @@ void onDataRehash(DataRehashedEvent dataRehashedEvent) {
                completableFutures.add(transferTask.requestSegments());
             }
          }
-         segmentRequestFuture = CompletableFuture
+         future = CompletableFuture
```

Comment:
I would also think about limiting the degree of parallelism here. I recall issues in 32 node tests, where a node joining the cluster requested state from all other nodes at once - and the other 31 nodes started sending it, effectively DDoSing the poor joiner.